### PR TITLE
Fallback to cloning repository

### PR DESCRIPTION
**Pull request by athos-ribeiro**
This is the first PR for the series of changes on handling corrupted archives.

When Cachito does not succeed on re-using an existing source archive to
retrieve the project's git history, try using other existing archives.
Additionally, if no archives can be used, fall back to cloning the
repository as it is done when an archive does not exist.

### CODE REVIEW

**Review by nirzari — APPROVED** [review]
LGTM

File: tests/test_workers/test_scm.py [hunk 1]
```diff
@@ -122,6 +123,42 @@ def test_fetch_source_by_pull(mock_update_and_archive, mock_getctime):
...
+    scm_git = scm.Git(url, ref)
+
+    po = mock.patch.object
+    with po(scm_git.sources_dir.archive_path, "exists", return_value=False):
+        with po(
+            scm_git.sources_dir.package_dir, "glob", return_value=["29eh2a.tar.gz", "a8c2d2.tar.gz"]
+        ):
+            scm_git.fetch_source()
+
+    mock_update_and_archive.call_count == 2
```
Comment (chmeliik) [inline tests/test_workers/test_scm.py:153]: Is the `assert` keyword missing here?

Comment (athos-ribeiro) [reply tests/test_workers/test_scm.py]: fixed

File: cachito/workers/scm.py [hunk 2]
```diff
@@ -133,14 +133,21 @@ def fetch_source(self):
...
-        else:
-            self.clone_and_archive()
+        for previous_archive in previous_archives:
+            try:
+                self.update_and_archive(previous_archive)
+                return
+            except (
+                git.exc.InvalidGitRepositoryError,
+                tarfile.ExtractError,
+                EnvironmentError,
```
Comment (chmeliik) [inline cachito/workers/scm.py:146]: I was confused  about `EnvironmentError` for a second, turns out on python3.3 and later it is just an alias of `OSError`: https://docs.python.org/3.8/library/exceptions.html#EnvironmentError

Should we use `OSError` here instead? Would it also be possible to add a comment explaining when an `OSError` might occur?

Comment (athos-ribeiro) [reply cachito/workers/scm.py:146]: I checked the wrong reference when looking for exceptions to capture here. I checked https://github.com/python/cpython/blob/2.7/Lib/tarfile.py when I wanted to check https://github.com/python/cpython/blob/3.7/Lib/tarfile.py, where the EnvironmentError occurrences were substituted for ether OSError or ExtractError ones.

I will change it to `OSError`

**Review by chmeliik — APPROVED** [review]
LGTM :+1:

**Pull request merged**